/**
 * Copyright (c) 2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef MERGE_ABC_ASSEMBLY_PROGRAM_H
#define MERGE_ABC_ASSEMBLY_PROGRAM_H

#include "assembly-program.h"
#include "assemblyRecord.h"
#include "assemblyFunction.h"
#include "assemblyLiterals.h"
#include "assemblyProgram.pb.h"
#include "arena_allocator.h"

namespace panda::proto {
class Program {
public:
    explicit Program()
        : allocator_(std::make_unique<panda::ArenaAllocator>(panda::SpaceType::SPACE_TYPE_COMPILER, nullptr, true))
    {
    }
    static void Serialize(const panda::pandasm::Program &program, proto_panda::Program &protoProgram);
    void Deserialize(const proto_panda::Program &protoProgram, panda::pandasm::Program &program);
private:
    std::unique_ptr<panda::ArenaAllocator> allocator_ {};
};
} // panda::proto
#endif